File: codes/python/chapter_tree/binary_tree_bfs.py
"""
File: binary_tree_bfs.py
Created Time: 2022-12-20
Author: a16su (lpluls001@gmail.com)
"""

import sys, os.path as osp
sys.path.append(osp.dirname(osp.dirname(osp.abspath(__file__))))
from modules import *


def level_order(root: Optional[TreeNode]) -> List[int]:
    """ 层序遍历 """
    # 初始化队列，加入根结点
    queue: Deque[TreeNode] = collections.deque()
    queue.append(root)
    # 初始化一个列表，用于保存遍历序列
    res: List[int] = []
    while queue:
        node: TreeNode = queue.popleft() # 队列出队
        res.append(node.val)             # 保存结点值
        if node.left is not None:
            queue.append(node.left)      # 左子结点入队
        if node.right is not None:
            queue.append(node.right)     # 右子结点入队
    return res


""" Driver Code """
if __name__ == "__main__":
    # 初始化二叉树
    # 这里借助了一个从数组直接生成二叉树的函数
    root: TreeNode = list_to_tree(arr=[1, 2, 3, 4, 5, 6, 7])
    print("\n初始化二叉树\n")
    print_tree(root)

    # 层序遍历
    res: List[int] = level_order(root)
    print("\n层序遍历的结点打印序列 = ", res)
    assert res == [1, 2, 3, 4, 5, 6, 7]

File: codes/lua/chapter_tree/binary_tree_bfs.lua
-- @script binary_tree_bfs.lua
-- @date 2025-11-13
-- @author fisheryv (yue.fisher2025@gdhfi.com)


-- 添加模块搜索路径
package.path = package.path .. ";codes/lua/modules/?.lua"

local print_util = require("print_util")
local tree_node = require("tree_node")
local list_to_tree = tree_node.list_to_tree

--- 层序遍历二叉树
--- @param root TreeNode|nil 二叉树的根节点
--- @return table
local function level_order(root)
    if not root then
        return {}
    end

    -- 初始化队列，加入根节点
    local queue = { root }
    -- 初始化一个列表，用于保存遍历序列
    local res = {}

    while #queue > 0 do
        -- 队列出队
        local node = table.remove(queue, 1)
        -- 保存节点值
        table.insert(res, node.val)

        -- 左子节点入队
        if node.left then
            table.insert(queue, node.left)
        end

        -- 右子节点入队
        if node.right then
            table.insert(queue, node.right)
        end
    end

    return res
end

-- Driver Code
local function main()
    -- 初始化二叉树
    -- 这里借助了一个从数组直接生成二叉树的函数
    local root = list_to_tree({ 1, 2, 3, 4, 5, 6, 7 })
    print("\n初始化二叉树\n")
    print_util.print_tree(root, nil, false)

    -- 层序遍历
    local res = level_order(root)
    print("\n层序遍历的节点打印序列 = [" .. table.concat(res, ", ") .. "]\n")
end

-- 执行主函数
main()
